### PR TITLE
C++: Nested field flow (without array content!)

### PR DESCRIPTION
**Pull request by MathiasVP**
(Fixes https://github.com/github/codeql-c-analysis-team/issues/280.)

This PR is _another_ attempt at getting better field flow for IR. Unlike the previous attempts, this PR avoids adding additional read/store steps with `ArrayContent`. In other versions of this PR, `ArrayContent` was used to step from an address to a `LoadInstruction` during field flow.

Instead, this PR steps through `LoadInstruction`s during field flow by using a new dataflow node called `AddressNodeRead`.

CPP-differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1930/, and another run to be sure that it's mostly noise: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1931/.

CPP-differences shows 6 lost results and 3 new results:
- The 3 new results all good genuine.
- The 6 lost results all seem to be false positives. These show up on `main` because we seem to add some surprising `ArrayContent`s around that part of the code on `vim`. This makes us go above the field flow limit and so we forget which exact field has actually been read and stored, and we end up conflating the `cmd` and the `arg` field in this scenario:
Screenshot 2021-03-29 at 13 13 04

@rdmarsh2 I'd appreciate your review on this one. Hopefully, this is the final iteration of this PR that you'll have to review! Let me know if you want to a chat on Zoom about any of it (I'll happily move some of my workday around to accommodate that).

### CODE REVIEW

**Comment by MathiasVP** [comment]
The failing tests should all be in the internal directory. I'll create an internal PR once this PR is ready to be merged.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:752]: I think there's a copy-paste mistake here. Is this just the steps up the address computation chain?

**Review by rdmarsh2 — COMMENTED** [review]
If I'm understanding this correctly, our `AccessPath`s won't distinguish between an access to a field of an object stored in a field  and an access to a field of an object that is pointed-to by a field, and confusion between those is prevented because a given `Field` has to be either a pointer/reference or a bare object. Do we need to worry about conflation between a field (say, `o.p`) and a dereference of that field? (`*(o.p)`)

**Comment by MathiasVP** [comment]
> If I'm understanding this correctly, our `AccessPath`s won't distinguish between an access to a field of an object stored in a field and an access to a field of an object that is pointed-to by a field, and confusion between those is prevented because a given `Field` has to be either a pointer/reference or a bare object.

That sounds correct. This is one of the crucial differences between this PR and the previous versions. We don't distinguish these two situations because `LoadInstruction`s is traversed without needing any `ArrayContent` https://github.com/github/codeql/pull/5510/files#diff-73001a3b1f9949f0a0dbbc5653e9620fccf5d376f8709e0c960d2551d68da849R700 (as this created too many array content read/stores).

> Do we need to worry about conflation between a field (say, o.p) and a dereference of that field? (*(o.p))

Good question. It looks like we now have _dataflow_ in this example:
```cpp
int* source();
void sink(int);

struct O {
  int* p;
  int a;
};

void set_p(O* o, int* x) { o->p = x; }

void test(O o) {
  set_p(&o, source());
  sink(*o.p);
}
```
where's this flow requires a taint step in the AST library (to get from `o.p` to `*o.p`). That's likely something we want to change, but I'm not totally sure how. Here is the relevant IR:

```
v21_8(void)           = Call[set_p]                  : func:r21_1, 0:r21_3, 1:r21_5
m21_9(unknown)        = ^CallSideEffect              : ~m21_7
m21_10(unknown)       = Chi                          : total:m21_7, partial:m21_9
v21_11(void)          = ^BufferReadSideEffect[0]     : &:r21_3, ~m20_6
v21_12(void)          = ^BufferReadSideEffect[1]     : &:r21_5, ~m21_10
m21_13(unknown)       = ^BufferMayWriteSideEffect[0] : &:r21_3
m21_14(O)             = Chi                          : total:m20_6, partial:m21_13
m21_15(unknown)       = ^BufferMayWriteSideEffect[1] : &:r21_5
m21_16(unknown)       = Chi                          : total:m21_10, partial:m21_15
r22_1(glval<unknown>) = FunctionAddress[sink]        :
r22_2(glval<O>)       = VariableAddress[o]           :
r22_3(glval<int *>)   = FieldAddress[p]              : r22_2
r22_4(int *)          = Load[?]                      : &:r22_3, ~m21_14
r22_5(int)            = Load[?]                      : &:r22_4, ~m21_16
v22_6(void)           = Call[sink]                   : func:r22_1, 0:r22_5
```

the reason we get flow all the way to the final load (i.e., `r22_5`) is because we have a `preReadStep` from `m21_14` to `r22_2`, and from there we have:
- A `readStep` from `AddressNodeRead(r22_2)` to `AddressNodeRead(r22_3)`
- Followed by a `addressFlowRead` step from `AddressNodeRead(r22_3)` to `AddressNodeRead(r22_4)`.

At this point we want to stop doing _dataflow_ and rely on an a taint step from `r22_4` to `r22_5`, but we actually get another `addressFlowRead` step from `AddressNodeRead(r22_4)` to `AddressNodeRead(r22_5)`, and ending in a `postReadStep` from `AddressNodeRead(r22_5)` to `r22_5`.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: It's not a copy-paste mistake per se. But I understand your confusion. The comment wasn't very clear. 788b5e012e654a7f50af07b560b7529fd59b3f37 adds more explanation (and an example) to the predicate now. Hopefully, that makes it more clear.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:717]: Good point. Fixed in 788b5e012e654a7f50af07b560b7529fd59b3f37.

**Comment by rdmarsh2** [comment]
Is there a `postReadStep` from `AddressNodeRead(r22_4)` to `r22_4`?

**Comment by rdmarsh2** [comment]
Notes from our call:
- The chain should stop at *each* `LoadInstruction`, with a `postReadStep` from the address operand to the result, so that taint can flow to `o.p` and then via the normal taint step to `*o.p`.
- Shortening the chains might also improve performance in building the address nodes.
- We don't need to do the `ArrayContent` at `LoadInstruction`s to make this particular flow work, but it may be worth investigating in the future.

**Comment by MathiasVP** [comment]
The conflation is fixed in b6d68fd39c13fe8ac45e7f420909c48e5e4242f8 following our conversation from yesterday @rdmarsh2 . I've started a new [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1966/) since this is a fairly big change. It looks like breaking up these chains is beneficial to the overall flow as well 🎉

**Review by rdmarsh2 — COMMENTED** [review]
Code changes look good otherwise.

I'm surprised that we lose partial definitions - is that from older commits that you hadn't updated the tests for?

**Comment by MathiasVP** [comment]
> I'm surprised that we lose partial definitions - is that from older commits that you hadn't updated the tests for?

I accidentally broke partial definitions in b6d68fd39c13fe8ac45e7f420909c48e5e4242f8. I've fixed them in 636da21965b6da340e9c7b00d69a2d82f99b9e33.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:629]: That'll never be the case in this disjunction as `nodeFrom` is an `InstructionNode` and `nodeTo` is an `AddressNodeRead`.

**Comment by MathiasVP** [comment]
> The conflation is fixed in [b6d68fd](https://github.com/github/codeql/commit/b6d68fd39c13fe8ac45e7f420909c48e5e4242f8) following our conversation from yesterday @rdmarsh2 . I've started a new [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1966/) since this is a fairly big change. It looks like breaking up these chains is beneficial to the overall flow as well 🎉.

Ouch:
time_split_absolute

and also ~1000 new query results 😮 There's gotta be something wrong. I'll investigate!

**Comment by MathiasVP** [comment]
The performance characteristics shown in the graph above has been fixed in 79acc23e620f1684f4f31dc5bff02011e27c997e (as shown by [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1987/)). This CPP-differences shows that we're back to the 3 new results / 6 removed results that we want.

A couple of performance bottlenecks were still present, however. Specifically:

The use of `IRBlock::getASuccessor+()` wasn't restricted sufficiently. This is fixed in 975c6474cfd3eefbcee4dd8fff5fd210facdf58b by replacing this transitive closure with a forward-backward algorithm similar to the way the dataflow library computes the transitive closure.

Finally, dff6dff0735555871b595c259b43268362986435 fixes a couple of bad join orders. I have no idea why my changes affect the join ordering of `frontier_phi_node`, but the `pragma` I inserted brings back the original join order. I have started a new CPP-differences to check the performance: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1988/.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:33]: This case isn't in the old definition. What's it used for? Read steps for an address computation that gets used as a function argument?

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:33]: It's really a refinement of the old restriction so that we generate fewer `TAddressNodeRead`. Previously, _any_ address that flowed to a `Load` or `ReadSideEffectInstruction` was given a `TAddressNodeRead` regardless of whether or not it ended up being used for read steps. But now we only include those addresses where it's possible to perform a read step. At first I also thought it would be enough to have:
```codeql
exists(Instruction read | usedForRead(read) |
  addressFlowInstrRTC(i, read)
}
```
which says that only those addresses that flow to a read step is given a `TAddressNodeRead`. But we also want to have a `TAddressNodeRead` for the conversion in this example:
```cpp
struct S {
  int x, y;
};

void test() {
  S s;
  sink(*(long*)&s.x);
}
```
where there are a couple of address steps that don't generate a read step:
```
r17_2(glval<S>)       = VariableAddress[s]    :
r17_3(glval<int>)     = FieldAddress[x]       : r17_2
r17_4(int *)          = CopyValue             : r17_3
r17_5(long *)         = Convert               : r17_4
r17_6(long)           = Load[?]               : &:r17_5, ~m16_2
```

here, we will have a read step from the `TAddressNodeRead(r17_2)` to `TAddressNodeRead(r17_3)`, but we're still not at the `LoadInstruction` (which is how we actually leave a sequence of `TAddressNodeRead`s and go back to flowing through `Instruction`s and `Operand`s).

So to actually ensure that `r17_4` and `r17_5` are assigned a `TAddressNodeRead` we also need to include the instructions that receive flow from a read step, which ends up looking like this:
```codeql
exists(Instruction read | usedForRead(read) |
  addressFlowInstrRTC(i, read)
  or
  addressFlowInstrRTC(read, i)
}
```

**Comment by MathiasVP** [comment]
> Edit: Locally, it looks like it's a slight improvement to replace the manual recursion in `IRBlockFlow` with a `fastTC`. I've started another (hopefully, final) CPP-differences to check this: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1992/.

CPP-differences look slightly better after using the `fastTC`. I don't think I need to tinker more with performance now.

**Review by rdmarsh2 — APPROVED** [review]

**Comment by rdmarsh2** [comment]
@jbj this looks like a 3-5% slowdown, depending on the codebase. I know we're expecting a performance hit because of the increased size of the data flow graph, but is that in the range we're willing to take?

**Comment by jbj** [comment]
> @jbj this looks like a 3-5% slowdown, depending on the codebase. I know we're expecting a performance hit because of the increased size of the data flow graph, but is that in the range we're willing to take?

The slowdown might be acceptable, but if this PR computes the use-use relationship between all pairs of uses, not just adjacent ones, then the inherent quadratic behaviour is not acceptable. It will surely explode on some codebase, either on lgtm.com or with a customer.

**Comment by MathiasVP** [comment]
> The slowdown might be acceptable, but if this PR computes the use-use relationship between all pairs of uses, not just adjacent ones, then the inherent quadratic behaviour is not acceptable. It will surely explode on some codebase, either on lgtm.com or with a customer.

I'll take another stab at avoiding the quadratic behavior in `postStoreStep` later today, then.

**Comment by MathiasVP** [comment]
> > The slowdown might be acceptable, but if this PR computes the use-use relationship between all pairs of uses, not just adjacent ones, then the inherent quadratic behaviour is not acceptable. It will surely explode on some codebase, either on lgtm.com or with a customer.
> 
> I'll take another stab at avoiding the quadratic behavior in `postStoreStep` later today, then.

Something that maybe wasn't clear when I discussed this with you, @jbj, is we don't add flow between _all_ use-use pairs. We add flow between a use of a `LoadOperand` that occur while computing an address for a `storeStep`, and the same `LoadOperand` that occur while computing the address for a `readStep`. That is, there's a flow step from `s.x` to both `s.x` and `s.y` in:
```cpp
argument_source(s.x);
sink(s.x);
sink(s.y);
```

AFAIK, this is the same strategy as with the `Node -> FlowVar -> VariableAccess` steps in the AST dataflow's `simpleLocalFlowStep` predicate: The partial definition of `s` in `argument_source(s.x)` flows to both of the variable accesses in `sink(s.x)` and `sink(x.y)`.

**Comment by jbj** [comment]
That sounds safer, but the use-use mechanism in AST data flow also has severe performance problems. In particular, it has this workaround that's been needed on several open-source codebases and at least one customer codebase:

https://github.com/github/codeql/blob/10e76ff28fc23509061491b6c2ddef75f9f06002/cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll#L616-L625

**Comment by MathiasVP** [comment]
Oh, right. That beast. Yeah, I'd for sure like to avoid doing the same for the IR. Good point!

**Comment by MathiasVP** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2009/ compares 51d1fba224b67354635bc04ed6ac60b692a18480 to the latest `main` that's been merged into this branch. The slowdown is still around 3-5%, and we still gain the 3 true positives and lose 6 false positives.

I think it would be good to run this on LGTM to see if there are any pathological projects.

**Comment by MathiasVP** [comment]
The LGTM results came back. The log files show that there are still problems related to `ArrayContent` on some projects.

**Comment by MathiasVP** [comment]
I'm closing this PR for now as my list of open PRs on https://github.com/pulls is about to overflow my tiny screen. We can revisit it once we start working on dataflow improvements again.

**Pull request closed without merging**
